src/components/ble/weather/WeatherService.cpp

```diff
@@ -42,315 +305,364 @@ namespace Pinetime {
         if (packetLen <= 0) {
           return BLE_ATT_ERR_INVALID_ATTR_VALUE_LEN;
         }
+
         // Decode
         QCBORDecodeContext decodeContext;
-        UsefulBufC encodedCbor = {ctxt->om->om_data, OS_MBUF_PKTLEN(ctxt->om)}; // NOLINT(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+        UsefulBufC encodedCbor = {ctxt->om->om_data, packetLen};
 
         QCBORDecode_Init(&decodeContext, encodedCbor, QCBOR_DECODE_MODE_NORMAL);
         // KINDLY provide us a fixed-length map
         QCBORDecode_EnterMap(&decodeContext, nullptr);
         // Always encodes to the smallest number of bytes based on the value
-        int64_t tmpTimestamp = 0;
-        QCBORDecode_GetInt64InMapSZ(&decodeContext, "Timestamp", &tmpTimestamp);
-        if (QCBORDecode_GetError(&decodeContext) != QCBOR_SUCCESS) {
+
+        auto optTimestamp = Get<uint64_t>(&decodeContext, "Timestamp");
+        if (!optTimestamp) {
           CleanUpQcbor(&decodeContext);
           return BLE_ATT_ERR_INVALID_ATTR_VALUE_LEN;
         }
-        int64_t tmpExpires = 0;
-        QCBORDecode_GetInt64InMapSZ(&decodeContext, "Expires", &tmpExpires);
-        if (QCBORDecode_GetError(&decodeContext) != QCBOR_SUCCESS || tmpExpires < 0 || tmpExpires > 4294967295) {
+
+        auto optExpires = Get<uint32_t>(&decodeContext, "Expires");
+        if (!optExpires) {
           CleanUpQcbor(&decodeContext);
           return BLE_ATT_ERR_INVALID_ATTR_VALUE_LEN;
         }
-        int64_t tmpEventType = 0;
-        QCBORDecode_GetInt64InMapSZ(&decodeContext, "EventType", &tmpEventType);
-        if (QCBORDecode_GetError(&decodeContext) != QCBOR_SUCCESS || tmpEventType < 0 ||
-            tmpEventType >= static_cast<int64_t>(WeatherData::eventtype::Length)) {
+
+        auto optEventType = Get<Pinetime::Controllers::WeatherData::eventtype>(&decodeContext, "EventType");
+        if (!optEventType) {
+          CleanUpQcbor(&decodeContext);
+          return BLE_ATT_ERR_INVALID_ATTR_VALUE_LEN;
+        }
+
+        // Do not add already expired events
+        uint64_t currentTimestamp = GetCurrentUnixTimestamp();
+        if (static_cast<uint64_t>(*optTimestamp + *optExpires) < currentTimestamp) {
+          CleanUpQcbor(&decodeContext);
+          return BLE_ATT_ERR_INVALID_ATTR_VALUE_LEN;
+        }
+
+        auto optID = Get<uint16_t>(&decodeContext, "EventID");
+        if (!optID) {
+          CleanUpQcbor(&decodeContext);
+          return BLE_ATT_ERR_INVALID_ATTR_VALUE_LEN;
+        }
+
+        std::unique_ptr<WeatherData::TimelineHeader>& existingEvent = GetEventByID(*optEventType, *optID);
+        bool useExisting = IsEventStillValid(existingEvent, currentTimestamp);
+        if (!useExisting && GetTimelineLength() >= maxNbElements) {
           CleanUpQcbor(&decodeContext);
           return BLE_ATT_ERR_INVALID_ATTR_VALUE_LEN;
         }
 
-        switch (static_cast<WeatherData::eventtype>(tmpEventType)) {
+        TidyTimeline();
```

Comment:
Not sure how long this call takes, but having accidentally injected some spin loops into BLE callbacks I can say that you can do processing for multiple seconds before returning a response

Reply:
I don't think this will be a problem. As Mark says, we can do quite a lot before needing to return a response, and I don't think that these operations will take very long anyway (looping through a maximum of 10 items and removing probably not many stale ones, and sorting them shouldn't take long). I have noticed that the check for how long the timeline is should probably be done after tidying the timeline, so I'll fix that and then mark this as resolved.